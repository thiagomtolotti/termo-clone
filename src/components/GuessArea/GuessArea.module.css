.guessArea {
  font-size: 72px;

  margin: auto;
}
