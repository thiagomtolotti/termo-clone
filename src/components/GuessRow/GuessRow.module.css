.guessRow {
  display: flex;
  gap: 4px;
  margin: 4px;
}

.guessRow.active {
  font-weight: 600;
}
