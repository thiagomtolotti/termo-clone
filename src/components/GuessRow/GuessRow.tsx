"use client";

import { useEffect, useState } from "react";
import { GuessLetter } from "../GuessLetter/GuessLetter";
import styles from "./GuessRow.module.css";

const WORD_SIZE = 5;
const letterRegex = /^[a-zA-Z]$/;

interface GuessRowProps {
  active?: boolean;
}

export const GuessRow = ({ active = false }: GuessRowProps) => {
  const [activeGuessIndex, setActiveGuessIndex] = useState(0);

  useEffect(() => {
    const handleArrowClick = (ev: KeyboardEvent) => {
      if (ev.key === "ArrowRight") {
        setActiveGuessIndex((activeIndex) =>
          Math.min(activeIndex + 1, WORD_SIZE - 1)
        );
      } else if (ev.key === "ArrowLeft") {
        setActiveGuessIndex((activeIndex) =>
          activeIndex !== null ? Math.max(activeIndex - 1, 0) : 0
        );
      }
    };
    const moveActiveLetterWhenLetterInput = (ev: KeyboardEvent) => {
      if (!letterRegex.test(ev.key)) return;

      setActiveGuessIndex((activeGuessIndex) =>
        Math.min(activeGuessIndex + 1, WORD_SIZE - 1)
      );
    };

    document.addEventListener("keydown", handleArrowClick);
    document.addEventListener("keydown", moveActiveLetterWhenLetterInput);

    return () => {
      document.removeEventListener("keydown", handleArrowClick);
      document.removeEventListener("keydown", moveActiveLetterWhenLetterInput);
    };
  }, [activeGuessIndex]);

  return (
    <div
      className={`${styles.guessRow} ${active ? styles.active : ""}`}
      role="guess-row"
    >
      {[...Array(WORD_SIZE)].map((_, index) => (
        <div
          onClick={() => {
            setActiveGuessIndex(index);
          }}
          role="guess-wrapper"
          key={index}
        >
          <GuessLetter active={active && activeGuessIndex === index} />
        </div>
      ))}
    </div>
  );
};
