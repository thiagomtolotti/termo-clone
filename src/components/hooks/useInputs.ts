import { useEffect, useLayoutEffect, useState } from "react";

const NUMBER_OF_GUESSES = 6;
const WORD_SIZE = 5;
const letterRegex = /^[a-zA-Z]$/;

export const useInputs = () => {
  const [currentRow, setCurrentRow] = useState(0);
  const [currentGuessIndex, setCurrentGuessIndex] = useState(0);
  const [rowsValue, setRowsValue] = useState(
    Array.from({ length: NUMBER_OF_GUESSES }, () => Array(WORD_SIZE).fill(""))
  );

  useEffect(() => {
    const handleLetterClick = (ev: KeyboardEvent) => {
      const keyPressed = ev.key;

      if (!letterRegex.test(keyPressed)) return;

      setRowsValue((rowsValue) => {
        const newValue = [...rowsValue];
        const activeRowValue = newValue[currentRow];

        activeRowValue[currentGuessIndex] = keyPressed.toUpperCase();

        return newValue;
      });

      setCurrentGuessIndex((currentGuessIndex) =>
        Math.min(currentGuessIndex + 1, WORD_SIZE - 1)
      );
    };

    const handleBackspaceClick = (ev: KeyboardEvent) => {
      const isBackspace = ev.key === "Backspace";

      if (!isBackspace) return;

      setRowsValue((rowValue) => {
        const newValue = [...rowValue];
        const activeRowValue = newValue[currentRow];

        activeRowValue[currentGuessIndex] = "";

        return newValue;
      });

      setCurrentGuessIndex((activeIndex) => Math.max(activeIndex - 1, 0));
    };

    const handleArrowClick = (ev: KeyboardEvent) => {
      const isArrowKey = ev.key === "ArrowLeft" || ev.key === "ArrowRight";

      if (!isArrowKey) return;

      if (ev.key === "ArrowRight") {
        setCurrentGuessIndex((activeIndex) =>
          Math.min(activeIndex + 1, WORD_SIZE - 1)
        );

        return;
      }

      setCurrentGuessIndex((activeIndex) => Math.max(activeIndex - 1, 0));
    };

    const handleEnterClick = (ev: KeyboardEvent) => {
      if (!ev.key.toLowerCase().includes("enter")) return;

      if (rowsValue[currentRow].indexOf("") !== -1) return;

      setCurrentRow((activeRow) => Math.min(activeRow + 1, NUMBER_OF_GUESSES));
    };

    document.addEventListener("keydown", handleLetterClick);
    document.addEventListener("keydown", handleBackspaceClick);
    document.addEventListener("keydown", handleArrowClick);
    document.addEventListener("keydown", handleEnterClick);

    return () => {
      document.removeEventListener("keydown", handleLetterClick);
      document.removeEventListener("keydown", handleBackspaceClick);
      document.removeEventListener("keydown", handleArrowClick);
      document.removeEventListener("keydown", handleEnterClick);
    };
  }, [currentRow, currentGuessIndex]);

  useLayoutEffect(() => {
    setCurrentGuessIndex(0);
  }, [currentRow]);

  return { rowsValue, currentRow, currentGuessIndex };
};
