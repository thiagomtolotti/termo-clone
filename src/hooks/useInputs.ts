import { useEffect, useLayoutEffect, useState } from "react";

const NUMBER_OF_GUESSES = 6;
const WORD_SIZE = 5;
const letterRegex = /^[a-zA-Z]$/;

export const useInputs = () => {
  const [rowsValue, setRowsValue] = useState(
    Array.from({ length: NUMBER_OF_GUESSES }, () => Array(WORD_SIZE).fill(""))
  );
  const [currentPosition, setCurrentPosition] = useState<number[]>([0, 0]);

  useEffect(() => {
    const handleLetterClick = (ev: KeyboardEvent) => {
      const keyPressed = ev.key;

      if (!letterRegex.test(keyPressed)) return;

      setRowsValue((rowsValue) => {
        const newValue = [...rowsValue];
        const activeRowValue = newValue[currentPosition[0]];

        activeRowValue[currentPosition[1]] = keyPressed.toUpperCase();

        return newValue;
      });

      setCurrentPosition((currentPosition) => {
        const [rowPos, colPos] = currentPosition;

        const newColPos = colPos + 1;

        return [rowPos, newColPos];
      });
    };

    const handleBackspaceClick = (ev: KeyboardEvent) => {
      const isBackspace = ev.key === "Backspace";

      if (!isBackspace) return;

      setRowsValue((rowValue) => {
        const newValue = [...rowValue];
        const activeRowValue = newValue[currentPosition[0]];

        activeRowValue[currentPosition[1]] = "";

        return newValue;
      });

      setCurrentPosition((currentPosition) => {
        const [rowPos, colPos] = currentPosition;

        const newColPos = colPos - 1;

        return [rowPos, newColPos];
      });
    };

    const handleArrowClick = (ev: KeyboardEvent) => {
      const isArrowKey = ev.key === "ArrowLeft" || ev.key === "ArrowRight";

      if (!isArrowKey) return;

      if (ev.key === "ArrowRight") {
        setCurrentPosition((currentPosition) => {
          const [rowPos, colPos] = currentPosition;

          const newColPos = colPos + 1;

          return [rowPos, newColPos];
        });

        return;
      }

      setCurrentPosition((currentPosition) => {
        const [rowPos, colPos] = currentPosition;

        const newColPos = colPos - 1;

        return [rowPos, newColPos];
      });
    };

    const handleEnterClick = (ev: KeyboardEvent) => {
      if (!ev.key.toLowerCase().includes("enter")) return;

      if (rowsValue[currentPosition[0]].indexOf("") !== -1) return;

      setCurrentPosition((currentPosition) => {
        const [rowPos, _] = currentPosition;

        const newRowPos = rowPos + 1;

        return [newRowPos, 0];
      });
    };

    document.addEventListener("keydown", handleLetterClick);
    document.addEventListener("keydown", handleBackspaceClick);
    document.addEventListener("keydown", handleArrowClick);
    document.addEventListener("keydown", handleEnterClick);

    return () => {
      document.removeEventListener("keydown", handleLetterClick);
      document.removeEventListener("keydown", handleBackspaceClick);
      document.removeEventListener("keydown", handleArrowClick);
      document.removeEventListener("keydown", handleEnterClick);
    };
  }, [currentPosition]);

  useLayoutEffect(() => {
    const activeGuess = currentPosition[1];
    const activeRow = currentPosition[0];

    if (activeGuess > WORD_SIZE - 1) {
      setCurrentPosition((currentPosition) => {
        const [rowPos, _] = currentPosition;

        const newColPos = WORD_SIZE - 1;

        return [rowPos, newColPos];
      });
    }

    if (activeGuess < 0) {
      setCurrentPosition((currentPosition) => {
        const [rowPos, _] = currentPosition;

        return [rowPos, 0];
      });
    }

    if (activeRow > NUMBER_OF_GUESSES - 1) {
      setCurrentPosition([-1, 0]);
    }
  }, [currentPosition]);

  // useLayoutEffect(() => {
  //   setCurrentPosition(currentPosition);
  // }, [currentRow]);

  return { rowsValue, currentPosition };
};
