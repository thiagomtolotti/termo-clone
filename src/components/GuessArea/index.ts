import { GuessArea } from "./GuessArea";

export default GuessArea;
