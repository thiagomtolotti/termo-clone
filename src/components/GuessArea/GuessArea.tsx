"use client";

import { useEffect, useState } from "react";
import { GuessRow } from "../GuessRow/GuessRow";

const NUMBER_OF_GUESSES = 6;

export const GuessArea = () => {
  const [activeRow, setActiveRow] = useState(0);

  useEffect(() => {
    const handleEnterClick = (ev: KeyboardEvent) => {
      if (!ev.key.toLowerCase().includes("enter")) return;

      setActiveRow((activeRow) =>
        Math.min(activeRow + 1, NUMBER_OF_GUESSES - 1)
      );
    };

    document.addEventListener("keydown", handleEnterClick);

    return () => {
      document.removeEventListener("keydown", handleEnterClick);
    };
  }, [activeRow]);

  return [...Array(NUMBER_OF_GUESSES)].map((_, index) => (
    <GuessRow active={index === activeRow} />
  ));
};
