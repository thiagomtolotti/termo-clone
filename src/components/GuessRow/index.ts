import { GuessRow } from "./GuessRow";

export default GuessRow;
