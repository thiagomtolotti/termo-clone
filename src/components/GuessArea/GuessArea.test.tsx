import { render } from "@testing-library/react";
import { GuessArea } from "./GuessArea";
import { GuessRow } from "../GuessRow/GuessRow";
import * as useInputsModule from "../../hooks/useInputs/useInputs";

jest.mock("../../hooks/useInputs/useInputs");

jest.mock("../GuessRow/GuessRow");

describe("GuessArea component", () => {
  it("Should render correctly", () => {
    const mockRowsValue = [
      ["T", "E", "S", "T", "E"],
      ["T", "E", "S", "T", "A"],
      ["T", "E", "S", "T", "B"],
      ["T", "E", "S", "T", "C"],
      ["T", "E", "S", "T", "D"],
      ["T", "E", "S", "T", "F"],
    ];
    const mockCurrentPosition = [1, 2];

    (useInputsModule.useInputs as jest.Mock).mockReturnValue({
      rowsValue: mockRowsValue,
      currentPosition: mockCurrentPosition,
    });

    const { container } = render(<GuessArea />);

    expect(container.innerHTML).toMatchSnapshot();
    expect(GuessRow).toHaveBeenCalledTimes(mockRowsValue.length);

    (GuessRow as jest.Mock).mock.calls.forEach((args, index) => {
      const [props] = args;

      expect(props.value).toBe(mockRowsValue[index]);

      if (index === mockCurrentPosition[0]) {
        expect(props.activeIndex).toBe(mockCurrentPosition[1]);
      } else {
        expect(props.activeIndex).toBe(undefined);
      }
    });
  });
});
