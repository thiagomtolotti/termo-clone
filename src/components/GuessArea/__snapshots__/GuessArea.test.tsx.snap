// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuessArea component Should render correctly 1`] = `"<div class="guessRow active" role="guess-row"><div><div role="guess" class="guess active"></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div></div><div class="guessRow " role="guess-row"><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div></div><div class="guessRow " role="guess-row"><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div></div><div class="guessRow " role="guess-row"><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div></div><div class="guessRow " role="guess-row"><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div></div><div class="guessRow " role="guess-row"><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div><div><div role="guess" class="guess "></div></div></div>"`;
